readthedocs/projects/tasks/utils.py

```diff
@@ -323,6 +323,160 @@ def deprecated_config_file_used_notification():
+        '<a href="https://blog.readthedocs.com/build-image-config-deprecated/">Read our blog post to migrate to "build.os"</a> '  # noqa
+        "and ensure your project continues building successfully."
+    )
+    failure_level = WARNING_PERSISTENT
+
+
+class DeprecatedBuildImageEmailNotification(Notification):
+    app_templates = "projects"
+    name = "deprecated_build_image_used"
+    subject = '[Action required] Update your ".readthedocs.yaml" file to use "build.os"'
+    level = REQUIREMENT
+
+    def send(self):
+        """Method overwritten to remove on-site backend."""
+        backend = EmailBackend(self.request)
+        backend.send(self)
+
+
+@app.task(queue="web")
+def deprecated_build_image_notification():
+    """
+    Send an email notification about using "build.image" to all maintainers of the project.
+
+    This is a scheduled task to be executed on the webs.
+    Note the code uses `.iterator` and `.only` to avoid killing the db with this query.
+    Besdies, it excludes projects with enough spam score to be skipped.
+    """
+    # Skip projects with a spam score bigger than this value.
+    # Currently, this gives us ~250k in total (from ~550k we have in our database)
+    spam_score = 300
+
+    projects = set()
+    start_datetime = datetime.datetime.now()
+    queryset = Project.objects.exclude(users__profile__banned=True)
+    if settings.ALLOW_PRIVATE_REPOS:
+        # Only send emails to active customers
+        queryset = queryset.filter(
+            organizations__stripe_subscription__status=SubscriptionStatus.active
+        )
+    else:
+        # Take into account spam score on community
+        queryset = queryset.annotate(spam_score=Sum("spam_rules__value")).filter(
+            Q(spam_score__lt=spam_score) | Q(is_spam=False)
+        )
+    queryset = queryset.only("slug", "default_version").order_by("id")
+    n_projects = queryset.count()
+
+    for i, project in enumerate(queryset.iterator()):
+        if i % 500 == 0:
+            log.info(
+                'Finding projects using "build.image" config key.',
+                progress=f"{i}/{n_projects}",
+                current_project_pk=project.pk,
+                current_project_slug=project.slug,
+                projects_found=len(projects),
+                time_elapsed=(datetime.datetime.now() - start_datetime).seconds,
+            )
+
+        # Only check for the default version because if the project is using tags
+        # they won't be able to update those and we will send them emails forever.
+        # We can update this query if we consider later.
+        version = (
+            project.versions.filter(slug=project.default_version).only("id").first()
+        )
+        if version:
+            # Use a fixed date here to avoid changing the date on each run
+            years_ago = timezone.datetime(2020, 8, 1)
```

Comment:
This is in the last 3 years, which feels like a long time? Is this what we want?

Reply:
I expanded to 3 years on purpose but I'm fine if we want to reduce it to 1.

On the config file deprecation we started with no restriction about dates and sent 65k emails only once. After that we limited to 1 year.

Here, with 3 years, I found around 16k projects, which I think it's fine.

Reply:
```suggestion
            years_ago = timezone.datetime(2022, 8, 1)
```